salt/proxy/cisconso.py

```diff
@@ -0,0 +1,351 @@
+would for a regular minion:
+
+.. code-block:: bash
+
+    salt-key -a <id you gave the cisconso host>
+
+You can confirm that the pillar data is in place for the proxy:
+
+.. code-block:: bash
+
+    salt <id> pillar.items
+
+And now you should be able to ping the Cisco NSO host to make sure it is
+responding:
+
+.. code-block:: bash
+
+    salt <id> test.ping
+'''
+
+# Import Python Libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt Libs
+from salt.exceptions import SaltSystemExit
+
+# This must be present or the Salt loader won't load this module.
+__proxyenabled__ = ['cisconso']
+
+try:
+    from pynso.client import NSOClient
+    from pynso.datastores import DatastoreType
+    HAS_PYNSO_LIBS = True
+except ImportError:
+    HAS_PYNSO_LIBS = False
+
+# Variables are scoped to this module so we can have persistent data
+# across calls to fns in here.
+GRAINS_CACHE = {}
+DETAILS = {}
+
+# Set up logging
+log = logging.getLogger(__file__)
+
+# Define the module's virtual name
+__virtualname__ = 'cisconso'
+
+
+def __virtual__():
+    return HAS_PYNSO_LIBS
+
+
+def init(opts):
+    # Set configuration details
+    DETAILS['host'] = opts['proxy'].get('host')
+    DETAILS['username'] = opts['proxy'].get('username')
+    DETAILS['password'] = opts['proxy'].get('password')
+    DETAILS['use_ssl'] = bool(opts['proxy'].get('use_ssl'))
+    DETAILS['port'] = int(opts['proxy'].get('port'))
+
+
+def grains():
+    '''
+    Get the grains from the proxy device.
+    '''
+    if not GRAINS_CACHE:
+        return _grains()
+    return GRAINS_CACHE
+
+
+def _get_client():
+    return NSOClient(
+        host=DETAILS['host'],
+        username=DETAILS['username'],
+        password=DETAILS['password'],
+        port=DETAILS['port'],
+        ssl=DETAILS['use_ssl'])
+
+
+def ping():
+    '''
+    Check to see if the host is responding. Returns False if the host didn't
+    respond, True otherwise.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt cisco-nso test.ping
+    '''
+    try:
+        client = _get_client()
+        client.info()
+    except SaltSystemExit as err:
+        log.warning(err)
+        return False
+
+    return True
+
+
+def shutdown():
+    '''
+    Shutdown the connection to the proxy device. For this proxy,
+    shutdown is a no-op.
+    '''
+    log.debug('Cisco NSO proxy shutdown() called...')
+
+
+def call_get_data(datastore, path):
+    '''
+    Get the configuration of the device tree at the given path
+
+    :param datastore: The datastore, e.g. running, operational.
+        One of the NETCONF store IETF types
+    :type  datastore: :class:`DatastoreType` (``str`` enum).
+
+    :param path: The device path, a list of element names in order,
+        comma seperated
+    :type  path: ``list`` of ``str`` OR ``tuple``
+
+    :return: The network configuration at that tree
+    :rtype: ``dict``
+
+    .. code-block:: bash
+
+        salt cisco-nso cisconso.get_data devices
```

Comment:
I was going to ask about that. I was comparing with the philips hue proxy minion and thought maybe `def call_x` was a convention to expose the execution action `module.x`?